added batch mailer

weilen-mc opened the pull request:
Added another mailer called batchmailer.py in the util folder. It allows emails with many (BCC) recipients to be sent in batches of size: EMAIL_BATCH_SIZE. The mailer waits between batches for a specified time: EMAIL_WAIT_BETWEEN_BATCHES. When the setting EMAIL_SEND_BATCH_ADMIN_MESSAGE is set to true the mailer will also send an email to the juntagrico admin with the recipient email adresses. No admin mail will be sent for emails containing less recipients than EMAIL_BATCH_SIZE.  

This mailer has been tested by us for several months and we had no more problems with spam filters (especially from bluewin) or hoster limits.

pascalfree commented on juntagrico/util/batchmailer.py:
```diff
@@ -0,0 +1,101 @@
+            # use default mail strategy if num of recipients is below batch size
+            if num_recipients <= email_batch_size:
+                msg.send()
+                batch = {'timestamp': str(timezone.now()),'to': tos, 'cc': ccs, 'bcc': bccs}
+                recipients_batches.append(batch)
```
these lines have no effect, the function returns in the next line, without using the variable.

pascalfree commented on juntagrico/util/batchmailer.py:
```diff
@@ -0,0 +1,101 @@
+
+    def send(msg):
+
+        def _send_batches(msg, email_batch_size, wait_time_batch):
```
Move this function outside of the Mailer class.

pascalfree commented on juntagrico/util/batchmailer.py:
```diff
@@ -0,0 +1,101 @@
+            if send_batch_email_admin_message:
+                _send_admin_msg(msg, recipients_batches)
+
+        def _send_admin_msg(msg, recipients_batches):
```
move function out of class

pascalfree commented on juntagrico/util/batchmailer.py:
```diff
@@ -0,0 +1,101 @@
+            admin_msg.body = admin_body
+            admin_msg.subject = f'[JUNTAGRICO] Email Recipients: {msg.subject}'
+
+            admins = getattr(settings, 'ADMINS', False)
+            if not admins:
+                return
+            admin_emails = []
+            for adm in admins:
+                email = adm[1]
+                if email is not None:
+                    admin_emails.append(email)
```
Use list comprehension:
```
admins = getattr(settings, 'ADMINS', [])
admin_emails = [adm[1] for adm in admins if adm[1]]
```
untested.

pascalfree commented on juntagrico/util/batchmailer.py:
```diff
@@ -0,0 +1,101 @@
+                if email is not None:
+                    admin_emails.append(email)
+
+            #print('Admin Emails')
```
remove commented out code (everywhere)

pascalfree reviewed (commented):
I left some comments that should help reduce complexity.
As we'll have 2 mailer classes, I would move them to util/mailer/batch and util/mailer/default (while keeping the defaultmailer.py for backwards compatibility)

Also the batch mailer should be added to the docs.

weilen-mc commented:
I refactored the code and moved the files to utils/mailer/ as per your suggestion. I still need to test the implementation on our instance though to verify that the sending of emails works as intended...

I touched some unrelated files to pass the flake8 verification and the subsequent tests produced an error... I don't think this error has anything to do with my changes imo, but maybe you could check?

weilen-mc commented:
So I tested the branch on our instance and the batchmail works as intended. How does the process look like to write the doc for this feature? Do I add it directly to the docs/1_5_release_notes.rst ?

pascalfree commented:
Regarding the summary emails for admins, I wonder how useful it is. Do you have this enabled in production?
In the end this only tells, if all emails have been sent. If the thread is interrupted for some reason, no summary will be sent to the admin and there is no way to know who already got the email. It is probably more useful to just print the summary (this can be viewed in the logs of the PaaS).

I fixed the test issue (it was unrelated to this PR).

pascalfree commented:
I refactored the batch mailer to reduce code duplication and add some features:
Inspired by the [ortoloco custom mailer](https://github.com/ortoloco/ortoloco/blob/live/ortoloco/mailer.py) the emails are sent with "to" recipients if the batch size is 1.
Also note that I moved the settings to our "config" object that we use in juntagrico and documented these settings.

I also added a test.

The admin notification I removed, but instead I'm now producing a signal, each time a batch email is sent. That way it is easy to add a function to react on these signals and e.g. send an admin notification. This would look something like this (in the models.py of your instance):

```
batch_mail_collector = {}


@receiver(batch_mail_sent)
def sent_batch(sender, message, original_message, done, **kwargs):
    if original_message not in batch_mail_collector:
        batch_mail_collector[original_message] = []
    if not done:
        batch_mail_collector[original_message].append(message.recipients())
    else:
        print(batch_mail_collector[original_message])  # or sent this in an email
        del batch_mail_collector[original_message]

```

Can you have a look @weilen-mc if that still fulfills your needs?

weilen-mc commented:
I quickly tested it on our instance and it works as intended as far as I could see. The changes are deployed and if an issue should pop up I'll report it here.
